src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.spec.ts

```diff
@@ -107,35 +107,28 @@ describe('OsdListComponent', () => {
   });
 
   describe('test table actions in submenu', () => {
+    let tableActionDebugElement;
+    let toClassName;
+
+    const getActionClasses = (action) =>
+      tableActionDebugElement.query(By.css('.' + toClassName(action.name))).classes;
+
     beforeEach(
       fakeAsync(() => {
         // The menu needs a click to render the dropdown!
         const dropDownToggle = fixture.debugElement.query(By.css('.dropdown-toggle'));
         dropDownToggle.triggerEventHandler('click', null);
         tick();
         fixture.detectChanges();
+        tableActionDebugElement = fixture.debugElement.query(By.directive(TableActionsComponent));
+        toClassName = TestBed.get(TableActionsComponent).toClassName;
```

Comment:
I assigned them in the before each because they were initiated first inside the loop of the test. Now they are only initiated once and not 9 times (9 actions).

Reply:
I think the point is about the scope. `toClassName` was only used once in the test and just appeared where it was used, namely the `test table actions in submenu` describe block. It was clear where the method belonged to. Other parts of the code couldn't call or access `toClassName` and they didn't need it. And they still don't need it. That's why I think it belongs in the block where it was before. I actually think the provided encapsulation of the blocks is one of the purposes of the describe blocks. If that function was used in other blocks, we could add it where you've added it. Or put two describe blocks into another block. Whatever makes sense in that situation.

If you're concerned about TestBed.**get**(TableActionComponent) being repeatedly called, I'd like to suggest you to use a variable `tableActionComponent` instead:
```typescript
    it('has menu entries disabled for entries without create permission', () => {
      const tableActionComponent = TestBed.get(TableActionsComponent);
      component.tableActions
        .filter((tableAction) => tableAction.permission !== 'create')
        .map((tableAction) => tableAction.name)
        .map(tableActionComponent.toClassName)
        .map((className) => getMenuItem(`.${className}`))
        .forEach((debugElement) => {
          expect(debugElement.classes.disabled).toBe(true);
        });
    });
```

Reply:
As this describes only has one test, no other tests can access it. But I could add some more tests to test if the right mark action is disabled or enabled depending on the osd state, which would need all the functions and the `beforeEach` as it is, too. This would make this describe not only testing the initial template state with no selection.

Reply:
@Devp00l just for my understanding:

> I assigned them in the before each because they were initiated first inside the loop of the test. Now they are only initiated once and not 9 times (9 actions).

I guess you comment relates to performance?

Reply:
I'm ok with extracting `toClassName` from the map, but since it is a "static" method, I think we can just define `let toClassName  = TestBed.get(TableActionsComponent).toClassName;` in the describe scope.
No need to assign it in a `beforeEach`.

Reply:
@tspmelo I thought it would fail because `TestBed` would not be available outside of test methods, but I tested it locally and it works! I will amend that.

@s0nea Yes because of the performance

Reply:
Okay, thanks for clarification here. From my point of view is performance < readability/cohesion for test code.

Reply:
@tspmelo I've tested it it works with the old test, and the test was green, but that was a false positive return, as no expect got run because the function was undefined. Additional I have run only the single `it`, if I run the `describe` surrounding `it`, it will fail with `Error: Cannot call Promise.then from within a sync test.`. That means the the only place to keep it is inside test methods as `TestBed` will be out of scope otherwise.

But I've found a way to do it without `TestBed` :+1: